### PR TITLE
Sphinx documentation

### LINKED ISSUE #44: Documentation
I think it'd be good to start making a proper documentation, ideally with API documentation.

I was thinking of (surprise) [copying h11's format](https://github.com/python-hyper/h11/blob/master/h11/_connection.py#L113-L124).

Thoughts?

### PR DESCRIPTION
**Pull request by yeraydiazdiaz**
Boilerplate Sphinx documentation, mostly a copy of the README plus completed docstrings from the code.

There are a few quirks with the way Sphinx deals with `typing.NamedTuple` classes.

Screenshot 2020-03-28 at 16 16 08

Notice how it adds the `property: Alias for field number 0`. I couldn't quite figure out how to fix that 😞 

Closes #44

### CODE REVIEW

File: docs/source/conf.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+#
+# import os
+# import sys
+# sys.path.insert(0, os.path.abspath('.'))
+
+
+# -- Project information -----------------------------------------------------
+
+project = "socksio"
+copyright = "2020, Seth Michael Larson"
```
Comment (sethmlarson) [inline docs/source/conf.py]: Can we change this to use `datetime.date.today().year` so we don't have to update it ever again? :)

File: docs/source/conf.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+author = "Seth Michael Larson"
+
+# The full version, including alpha/beta/rc tags
+release = "0.2.0"
```
Comment (sethmlarson) [inline docs/source/conf.py]: Let's add an import and grab `socksio.__version__`.

File: docs/source/index.rst [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+========================================================
+
+``socksio`` is a sans-I/O library similar to
+`h11 <https://github.com/python-hyper/h11>`_ or
+`h2 <https://github.com/python-hyper/hyper-h2/>`_, this means the
+library itself does not handle the actual sending of the bytes through
+the network, it only deals with the implementation details of the SOCKS
+protocols so you can use it in any I/O library you want.
+
+Current status: alpha
```
Comment (sethmlarson) [inline docs/source/index.rst]: We can upgrade our status to stable I think, we're pretty confident with this API (as log as you agree!)
If you do agree maybe it's time for v1 :tada:

Can always release a breaking change with semver.

File: docs/source/api_socks4.rst [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+SOCKS4 and SOCKS4A API documentation
+====================================
+
+SOCKS4 and SOCKS4A are almost identical protocols, as such the API is implemented
+in a single module and most components are shared.
+
+The only practical difference is the usage of a :class:`SOCKS4Request` versus
+:class:`SOCKS4Request`.
+
+Remember SOCKS4 allows only for IPv4 addresses, i.e. no IPv6 or domain names.
```
Comment (sethmlarson) [inline docs/source/api_socks4.rst]: For SOCKS4 we should mention that SOCKS4a can handle domain names.

File: docs/source/conf.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
+# Configuration file for the Sphinx documentation builder.
```
Comment (sethmlarson) [inline docs/source/conf.py]: Let's point our linter at this file too

**Comment by sethmlarson** [comment]
Btw added you as owner on PyPI and sent you a message on Keybase :)

File: docs/source/index.rst (hunk 3 above)
Comment (yeraydiazdiaz) [reply docs/source/index.rst]: Agreed! 🚀

**Comment by yeraydiazdiaz** [comment]
> Also regarding the weird documentation behavior for the named tuples, maybe we need to document all their parameters in the class and then be picky about which members are documented (ie manually specify from_address() to be documented and that's it?
Would be a one-time thing and would look a lot cleaner.

Yep, still a bit annoying these quirks prevent using autodoc properly but this way we can control the ordering as well, I wanted to place the connection objects at the top.

> Is there a way we can add this to CI?

Sure thing, I guess we'd ideally have it build and push to read the docs on merged PRs?

**Comment by sethmlarson** [comment]
I think I've already got that setup for RTD :) I'll verify it works after we merge!

**Pull request merged**